### PR TITLE
Change Financial Data UI and session data persistence

### PR DESCRIPTION
**Pull request by WolfyUK**
### Context
[AB#208014](https://dfe-ssp.visualstudio.com/a14e55df-4fbf-4a2f-a11d-22b187178343/_workitems/edit/208014) [AB#198080](https://dfe-ssp.visualstudio.com/a14e55df-4fbf-4a2f-a11d-22b187178343/_workitems/edit/198080)

### Change proposed in this pull request
Populate the form with accordions of input fields on the Financial Data view. Implement Post action with model validation and PRG pattern for failed validation journeys.

### Guidance to review 
Browsing to Custom Data landing page and logging in displays the first form for custom data capture. Each input field should only accept decimals >=0 and provide validation message and summary if not. If successful, values persisted to session storage so if you leave and return those values should be pre-populated alongside the 'current' figures. Successfully validated forms will move onto the next stage of the journey as part of a subsequent work item.

> **NOTE:** Integration tests for the entire journey will be covered by a separate ticket

### Checklist
- [x] Work items have been linked [(use AB#)](https://learn.microsoft.com/en-us/azure/devops/boards/github/link-to-from-github?view=azure-devops#use-ab-to-link-from-github-to-azure-boards-work-items)
- [x] Your code builds clean without any errors or warnings
- [x] You have run all unit/integration tests and they pass
- [x] Your branch has been rebased onto main
- [x] You have tested by running locally

### CODE REVIEW

File: web/src/Web.App/Controllers/SchoolCustomDataChangeController.cs [hunk 1]
```diff
@@ -50,4 +47,38 @@ public async Task<IActionResult> Index(string urn)
...
+    public IActionResult FinancialData(string urn, [FromForm] SchoolCustomDataViewModel viewModel)
+    {
+        using (logger.BeginScope(new { urn, viewModel }))
+        {
+            try
+            {
+                if (!ModelState.IsValid)
+                {
+                    logger.LogDebug("Posted FinancialData failed validation: {ModelState}",
+                        JsonSerializer.Serialize(ModelState.Where(m => m.Value != null && m.Value.Errors.Any())));
```
Comment (jrabbott) [inline web/src/Web.App/Controllers/SchoolCustomDataChangeController.cs:63]: We have extension method for serialize and deserialize, to apply standard settings and formatting

File: web/src/Web.App/Attributes/NumericValueAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Web.App.Attributes;
+
+public class NumericValueAttribute : RegularExpressionAttribute
+{
+    public NumericValueAttribute() : base(@"^\s*-{0,1}((\d{1,}\.\d{1,})|(\d*))\s*$")
+    {
+        ErrorMessage = "Please enter numerical values only for {0}";
```
[the remark this answers is not in the document]
Comment (jrabbott) [reply web/src/Web.App/Attributes/NumericValueAttribute.cs:9]: Ticket needs updating either with a reason why we aren't following GDS or with updated text.

File: web/src/Web.App/Program.cs [hunk 3]
```diff
@@ -20,10 +20,17 @@
...
 builder.Host.UseSerilog((context, configuration) => configuration.ReadFrom.Configuration(context.Configuration));
 
-builder.Services.AddControllersWithViews().AddNewtonsoftJson(options =>
-{
-    options.SerializerSettings.SetJsonOptions();
-});
+builder.Services.AddControllersWithViews()
+    .AddNewtonsoftJson(options =>
+    {
+        options.SerializerSettings.SetJsonOptions();
```
[the remark this answers is not in the document]
Comment (WolfyUK) [reply web/src/Web.App/Program.cs:26]: This line was a formatting change only. Do we want to remove `SetJsonOptions()`?

[the remark this answers is not in the document]
Comment (jrabbott) [reply web/src/Web.App/Program.cs:26]: Oh yeah... sorry missed that, and also my bad I didn't read the code fully.

SetJsonOptions is actually part of the extension methods class, to set the standard options.

File: web/src/Web.App/Program.cs [hunk 4]
```diff
@@ -20,10 +20,17 @@
...
+    })
+    .AddMvcOptions(options =>
+    {
+        options.ModelBindingMessageProvider.SetAttemptedValueIsInvalidAccessor((_, field) =>
+            $"Please enter a valid value for {field}");
```
[the remark this answers is not in the document]
Comment (WolfyUK) [reply web/src/Web.App/Program.cs:31]: Resolved in #728

File: web/src/Web.App/Attributes/PositiveNumericValueAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Web.App.Attributes;
+
+public class PositiveNumericValueAttribute : RangeAttribute
+{
+    public PositiveNumericValueAttribute() : base(0, double.MaxValue)
+    {
+        ErrorMessage = "Please enter positive values only for {0}";
```
[the remark this answers is not in the document]
Comment (WolfyUK) [reply web/src/Web.App/Attributes/PositiveNumericValueAttribute.cs:9]: Resolved in #728

File: web/src/Web.App/Attributes/NumericValueAttribute.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (WolfyUK) [reply web/src/Web.App/Attributes/NumericValueAttribute.cs:9]: Resolved in #728

**Pull request merged**